### DISCUSSION TITLE
PWA support brainstorm

### DISCUSSION THREAD

**Discussion by ianthetechie** [stadiamaps/ferrostar#49] [Engineering RFDs]
Preface: I've been a mobile developer since 2008. I've tried a few PWAs over the years and I think they universally offer a worse experience than a native app and are usually of inferior quality. That said, I'm trying to keep an open mind ;)

I am seeing an increased amount of noise around the space from entities like the European Commission, on social media, and while I disagree with most of the takes, I want to have a discussion thread here so that Ferrostar can support these use cases if they become viable in the future. Knowing my own inability to predict trends and the tendency of worse tech that's easier to adopt to gain a foothold, I expect this will happen at some point so I want to prepare for it.

What we need to do for Ferrostar short term: support compilation to wasm targets and be ready to offer good JS bindings. We are already most of the way there do to the explicit design choices around the core. **We already support builds targeting WASI today, though it's not explicitly tested or advertised yet.**

What I'd like input from the community on (in no particular order):

* Can anyone point me to examples of apps or to documentation demonstrating that it is possible to get location updates with enough frequency to actually be usable for navigation?
* Can PWAs offer persistent notifications that most users would expect from a navigation experience? I believe they CAN generate local notifications but I'm not sure it's fully on par with native.
* I assume there is no possibility of CarPlay or Android auto today, but maybe this changes in the future.
* Can this be used in the background or does the app have to be active?

**Comment by ianthetechie** [comment]
Quick bit of research to drop here before I forget how to find it. At the moment, _background_ location support for PWAs is [currently nonexistent](https://github.com/w3c/geolocation-sensor/issues/22). This necessarily limits what could be built as things currently stand. However, we _can_ [watch updates in the foreground](https://w3c.github.io/geolocation-api/#watchposition-method).

**Comment by ianthetechie** [comment]
[This site](https://whatpwacando.today/) also has some broad info, but it's actually not very deep. Looks like this _is_ broadly possible with some limitations. These may still be valuable for things like a local government, tourism, etc. which don't have tech budget to hire a team of native devs, but they want to provide some level of navigation (ex: pedestrian or low-speed vehicle).

**Comment by Ali7040** [comment]
@ianthetechie  Hi, I searched on this topic, and here are the approaches I found.
1. **First Approach: Using Google Web API (Geolocation)**
   - Utilizing the Google Web API called Geolocation allows us to access the user's location with their permission. However, it cannot be used for full-fledged navigation experiences in PWAs due to the lack of background location access.

2. **Second Approach: Hybrid Approach**
   - Hybrid approaches, combining aspects of PWAs and native apps, can be a viable option for navigation. This approach leverages native capabilities for features like background location and richer notifications, potentially enabling full-fledged navigation experiences.
   - It provides access to device features like GPS and sensors not readily available in traditional PWAs.
   
**Drawbacks:**
   - Requires separate codebases for native platforms, adding complexity compared to pure PWAs.
   - May have larger app sizes compared to PWAs due to additional native code.

**Comment by ianthetechie** [comment]
Closing this discussion as it's been implemented :) https://stadiamaps.github.io/ferrostar/web-getting-started.html
